### PR TITLE
Xiaomi-Onoff-Switch

### PR DESCRIPTION
**Pull request by xrust83**
Will this work?

### CODE REVIEW

**Comment by madchicken** [comment]
It should, did you try it?

**Comment by xrust83** [comment]
No, I'm not experienced enough to compile this or create a repository, but when I made edits I looked at the analogy on other buttons

**Comment by madchicken** [comment]
Ok, I'll try it locally in some way. Give me some time, I am super busy in this period with my real job 😬

**Comment by wamfox** [comment]
I have tried to compile and run this code myself but yarn throws some errors I have attempted to fix this myself and have gotten it to compile but my button is unresponsive. I will add some comments to this code soon.

File: src/accessories/xiaomi/xiaomi-wireless-switch.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      .withStatelessSwitch('ON', 'on', 1, [
+        ProgrammableSwitchEvent.SINGLE_PRESS,
+        ProgrammableSwitchEvent.DOUBLE_PRESS,
+        ProgrammableSwitchEvent.LONG_PRESS,
+      ])
+      .andBattery()
+      .build();
+
+    return [
+      this.switchServiceSinglePress,
```
Comment (wamfox) [inline src/accessories/xiaomi/xiaomi-wireless-switch.ts:33]: I don't believe 'switchServiceSinglePress' is an available service I think this should be 'switchServiceToggle'.

File: src/accessories/xiaomi/xiaomi-wireless-switch.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+      this.platform,
+      this.accessory,
+      this.client,
+      this.state
+    );
+
+    const ProgrammableSwitchEvent = this.platform.Characteristic.ProgrammableSwitchEvent;
+  
+[
+  this.switchServicePress,
```
Comment (wamfox) [inline src/accessories/xiaomi/xiaomi-wireless-switch.ts:22]: I don't believe 'switchServicePress' is an available service I think this should be 'switchServiceToggle'.

File: src/accessories/xiaomi/xiaomi-wireless-switch.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      this.batteryService,
+    ];
+  }
+
+  update(device: Device, state: DeviceState) {
+    const ProgrammableSwitchEvent = this.platform.Characteristic.ProgrammableSwitchEvent;
+    super.update(device, state);
+    switch (state.click) {
+      case 'single_click':
+        this.switchServicePress
```
Comment (wamfox) [inline src/accessories/xiaomi/xiaomi-wireless-switch.ts:43]: I don't believe 'switchServicePress' is an available service I think this should be 'switchServiceToggle'.

File: src/accessories/xiaomi/xiaomi-wireless-switch.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+          .getCharacteristic(ProgrammableSwitchEvent)
+          .setValue(ProgrammableSwitchEvent.SINGLE_PRESS);
+        break;
+      case 'double_click':
+        this.switchServicePress
```
Comment (wamfox) [inline src/accessories/xiaomi/xiaomi-wireless-switch.ts:48]: I don't believe 'switchServicePress' is an available service I think this should be 'switchServiceToggle'.

File: src/accessories/xiaomi/xiaomi-wireless-switch.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+          .getCharacteristic(ProgrammableSwitchEvent)
+          .setValue(ProgrammableSwitchEvent.DOUBLE_PRESS);
+        break;
+      case 'hold_click':
+        this.switchServicePress
```
Comment (wamfox) [inline src/accessories/xiaomi/xiaomi-wireless-switch.ts:53]: I don't believe 'switchServicePress' is an available service I think this should be 'switchServiceToggle'.

File: src/accessories/xiaomi/xiaomi-wireless-switch.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+      this.client,
+      this.state
+    );
+
+    const ProgrammableSwitchEvent = this.platform.Characteristic.ProgrammableSwitchEvent;
+  
+[
+  this.switchServicePress, 
+  this.batteryService] = builder
+      .withStatelessSwitch('ON', 'on', 1, [
```
Comment (wamfox) [inline src/accessories/xiaomi/xiaomi-wireless-switch.ts:24]: I think the stateless switch should look something like this ('ON/OFF', 'toggle', 1,

File: src/accessories/xiaomi/xiaomi-wireless-switch.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
+import { ZigBeeAccessory } from '../zig-bee-accessory';
+import { Service } from 'homebridge';
+import { DeviceState } from '../../zigbee/types';
+import { Device } from 'zigbee-herdsman/dist/controller/model';
+import { ProgrammableSwitchServiceBuilder } from '../../builders/programmable-switch-service-builder';
+
+export class XiaomiWirelessSwitch extends ZigBeeAccessory {
+  protected switchServicePress: Service;
```
Comment (wamfox) [inline src/accessories/xiaomi/xiaomi-wireless-switch.ts:8]: I don't believe 'switchServicePress' is an available service I think this should be 'switchServiceToggle'. There are a few places where this needs to be changed

**Review by wamfox — COMMENTED** [review]
Have commented in my changes. It compiles for me but the button is unresponsive

File: src/accessories/xiaomi/xiaomi-wireless-switch.ts (hunk 1 above)
Comment (wamfox) [reply src/accessories/xiaomi/xiaomi-wireless-switch.ts:33]: sorry don't really know how to use Git this is a bit spamy

**Comment by wamfox** [comment]
This is my current '[xiaomi-wireless-switch.ts](https://github.com/wamfox/homebridge-zigbee-nt/blob/master/src/accessories/xiaomi/xiaomi-wireless-switch.ts)' file

after device paring, I do not get the device successfully configured in the log after atempting to mount

**Comment by madchicken** [comment]
> there doesnt seem to be any info in the device payload

This is weird. We probably need to check the zigbee-herdsman lib about this. Is it the device fully supported by zigbee2mqtt?

**Comment by wamfox** [comment]
I believe so but I'm not sure, I don't understand how to check what is and isn't in the herdsman library. It is present in the [supported devices for Zigbee2MQTT ](https://www.zigbee2mqtt.io/devices/WXKG01LM.html) however.

File: src/accessories/xiaomi/xiaomi-wireless-switch.ts (hunk 6 above)
Comment (xrust83) [reply src/accessories/xiaomi/xiaomi-wireless-switch.ts:24]: isn't toggle a commit key that responds to a state change?  the button in question has the function of clicking.  regular click, double click, long click.

File: src/accessories/xiaomi/xiaomi-wireless-switch.ts (hunk 7 above)
Comment (xrust83) [reply src/accessories/xiaomi/xiaomi-wireless-switch.ts:8]: the toggle service is used for tumblers that remain on or off after a normal light switch.  aqara button has click functions

File: src/accessories/xiaomi/xiaomi-wireless-switch.ts (hunk 6 above)
Comment (wamfox) [reply src/accessories/xiaomi/xiaomi-wireless-switch.ts:24]: You could be very correct I was basing this of the already implemented [Ikea remote](https://github.com/madchicken/homebridge-zigbee-nt/blob/master/src/accessories/ikea/ikea-remote-switch.ts) that uses a toggle for its middle button.

**Pull request closed without merging**